xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -300,15 +412,74 @@ public void run() {
     }
   }
 
-  // https://github.com/google/error-prone/issues/1767
-  @SuppressWarnings("ModifyCollectionInEnhancedForLoop")
-  private class ConfigWatcherImpl implements ConfigWatcher {
+  private class ResolveState implements LdsResourceWatcher {
+    private final ConfigOrError emptyServiceConfig =
+        serviceConfigParser.parseServiceConfig(Collections.<String, Object>emptyMap());
+    private final ResolutionResult emptyResult =
+        ResolutionResult.newBuilder()
+            .setServiceConfig(emptyServiceConfig)
+            // let channel take action for no config selector
+            .build();
     private Set<String> existingClusters;
+    @Nullable
+    private String rdsResource;
+    @Nullable
+    private RdsResourceWatcher rdsWatcher;
+
+    @Override
+    public void onChanged(LdsUpdate update) {
+      logger.log(XdsLogLevel.INFO, "Receive LDS resource update: {0}", update);
+      List<VirtualHost> virtualHosts = update.getVirtualHosts();
+      String rdsName = update.getRdsName();
+      if (rdsName != null && rdsName.equals(rdsResource)) {
+        return;
+      }
+      cleanUpRdsWatcher();
+      if (virtualHosts != null) {
+        updateRoutes(virtualHosts);
+      } else {
+        rdsResource = rdsName;
+        rdsWatcher = new RdsResourceWatcherImpl();
+        logger.log(XdsLogLevel.INFO, "Start watching RDS resource {0}", rdsResource);
+        xdsClient.watchRdsResource(rdsResource, rdsWatcher);
+      }
+    }
+
+    @Override
+    public void onError(Status error) {
+      logger.log(
+          XdsLogLevel.WARNING,
+          "Received error from xDS client {0}: {1}", xdsClient, error.getDescription());
+      listener.onError(error);
+    }
 
     @Override
-    public void onConfigChanged(ConfigUpdate update) {
+    public void onResourceDoesNotExist(String resourceName) {
+      logger.log(XdsLogLevel.INFO, "LDS resource {0} unavailable", resourceName);
+      cleanUpRdsWatcher();
+      listener.onResult(emptyResult);
+    }
+
+    private void start() {
+      logger.log(XdsLogLevel.INFO, "Start watching LDS resource {0}", authority);
+      xdsClient.watchLdsResource(authority, this);
+    }
+
+    private void stop() {
+      logger.log(XdsLogLevel.INFO, "Stop watching LDS resource {0}", authority);
+      xdsClient.cancelLdsResourceWatch(authority, this);
+      cleanUpRdsWatcher();
```

Comment:
Should they be cleaned up in a FILO order?

Reply:
It shouldn't matter. But sure, swapped.